### PR TITLE
[Diagnostics] Improve error when type parameters aren't equal

### PR DESCRIPTION
**Pull request by mdiep**
This improves the error for code like the following:

```swift
var a = [1,2,3,4]
var b = a.popFirst()
```

Before it would give this error:

```
error: '[Int]' requires the types '[Int]' and 'ArraySlice<Int>' be equivalent to use 'popFirst'
var b = a.popFirst()
          ^
```

I found that to be very unclear about why that was a requirement.

Now it gives this error, with note about where the method was declared:

```
error: 'Collection' requires the types '[Int]' and 'Array<Int>.SubSequence' ('ArraySlice<Int>') be equivalent to use 'popFirst'
var b = a.popFirst()
          ^
Swift.Collection:2:37: note: 'popFirst()' declared here
    @inlinable public mutating func popFirst() -> Self.Element?
                                    ^
```

That seems much clearer to me.

I'm not sure if it's possible to use a locally-built copy of Swift in Xcode, so I'm not sure if that location is enough for tools like Xcode to properly link to the declaration of `popFirst`.

Resolves [SR-7177](https://bugs.swift.org/browse/SR-7177).

### CODE REVIEW

File: include/swift/AST/DiagnosticsSema.def [hunk 1]
```diff
@@ -1432,6 +1432,9 @@ ERROR(types_not_equal,none,
 ERROR(types_not_equal_in_call,none,
       "%0 requires the types %1 and %2 be equivalent to use %3",
       (Type, Type, Type, DeclName))
+ERROR(types_not_equal_in_call_dependent,none,
```
Comment (mdiep) [inline include/swift/AST/DiagnosticsSema.def]: This ID is not my best naming, but nothing better jumped out at me.

File: lib/Sema/CSDiag.cpp [hunk 2]
```diff
@@ -1757,8 +1757,30 @@ static bool diagnoseTypeRequirementFailure(ConstraintSystem &cs,
...
+          }
+        } else {
+          TC.diagnose(UDE->getLoc(), diag::types_not_equal_in_call, baseType, lhs,
+                      rhs, UDE->getName());
+        }
+        TC.diagnose(member, diag::decl_declared_here, member->getFullName());
+      } else {
+        TC.diagnose(UDE->getLoc(), diag::types_not_equal_in_call, ownerType, lhs,
+                    rhs, UDE->getName());
+      }
```
Comment (mdiep) [inline lib/Sema/CSDiag.cpp]: This is… not as clean as I'd like. I'm very open to suggestions about how it might be improved.

Comment (xedin) [reply lib/Sema/CSDiag.cpp]: At least to me new messages don't seem to be much clearer about what is going on especially  for`.popFirst()`, maybe a better solution here would be to print the actual requirement which got broken like we do in other cases e.g. https://github.com/apple/swift/blob/master/test/Constraints/generics.swift#L251, WDYT?

Comment (mdiep) [reply lib/Sema/CSDiag.cpp]: Hmm… it does seem like it be nice to have parity there 🤔 

Unfortunately, I'm not sure it's possible without a somewhat major change to this diagnostic. (I'd love to be wrong on this.)

That information is available on the `Requirement`. I can get to that through `member`'s `GenericContext`. It looks like the `Constraint` is generated from the `Requirement`, but I don't see a way to connect that `Requirement` back to the current `Constraint` to know which of the `Requirement`s failed.

I tried moving the `diagnoseArgumentGenericRequirements` earlier—to be before this check—but it didn't hit.

It sure seems like there ought to be a way to unify these two things. But there's quite a bit of code between them, and I don't understand enough of the big picture to know how to connect them.

This PR seems like an improvement to me, so I'd probably take this improvement for now. But I'm happy to spend some more time looking—especially if you have any suggestions.

Comment (xedin) [reply lib/Sema/CSDiag.cpp]: Constraint locator should tell you exactly what requirement is this constraint related to and you can get it by index from generic signature of the context, take a look at how `open{Unbound}Generic`/`openGenericRequirements` generates such constraints.

Comment (mdiep) [reply lib/Sema/CSDiag.cpp]: How does this look?

```
error: value of type '[Int]' has no member 'popFirst'
var b = a.popFirst()
          ^
Swift.Collection:2:37: note: candidate requires that the types '[Int]' and 'ArraySlice<Int>' be equivalent (requirement specified as 'Self' == 'Self.SubSequence')
    @inlinable public mutating func popFirst() -> Self.Element?
                                    ^
```

Comment (xedin) [reply lib/Sema/CSDiag.cpp]: @mdiep I'm not sure what is going on but for some reason I can't see any new changes... But the new node looks great though!

**Comment by mdiep** [comment]
Okay, just pushed the updated code and tests.

Thanks for the help on this one, @xedin!

**Review by xedin — APPROVED** [review]
LGTM! One last nit request - could you please use `clang-format` on the commit before we merge this?

**Comment by mdiep** [comment]
> could you please use clang-format on the commit before we merge this?

Done!

**Comment by xedin** [comment]
@swift-ci please test and merge

**Comment by xedin** [comment]
@swift-ci please test and merge

**Comment by DougGregor** [comment]
This looks fantastic, thank you!

**Pull request merged**